Ai add multiple resource support

ron-murhammer opened the pull request:
AI - Updated purchase code to check all resources not just PUs. This prevents purchase failures due to not having enough non-PU resources that previously were occurring. Purchase logic still needs added to 'value' other resources as these changes only check for them.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/triplea-game/triplea/205)
<!-- Reviewable:end -->

veqryn commented:
The AI code is fairly self contained, so no need to assign me to review.
The is the one package in the project where I am fine with you just merging your stuff without review, though you are welcome to get additional feedback from Dan and others.

DanVanAtta commented:
Not a fan of changes that will conflict with #149, would be the 3rd time that PR has merge conflicts while waiting (meaning we duplicating work already done, and creating more work in the form of conflict resolution)

DanVanAtta commented on src/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java:
```diff
@@ -124,21 +125,25 @@ public boolean canUnitsBePlaced(final List<Unit> units, final PlayerID player, f
       final List<Unit> unitsToPlace, final Map<Territory, ProPurchaseTerritory> purchaseTerritories) {
+
     for (final Iterator<ProPurchaseOption> it = purchaseOptions.iterator(); it.hasNext();) {
       final ProPurchaseOption ppo = it.next();
```
Short variable names like "ppo" are really only appropriate for really short methods. Even then favor a variable name that clearly conveys its intent and pupose. Please write this variable name in long form to make it easier to understnad its usage later in the code.

DanVanAtta commented:
Review status: 0 of 5 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1210 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Ly-vxN5FNrcdY6JPV)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1210)):</sup>
Looks like resourcedata is only really used in this method. Can it be passed as a parameter rather than being a data member? The extra state we keep track of makes this class complicated. For example, could you have a NPE here? It's somewhat hard to track down and figure out. It looks like nothing prevents the resourceData that is passed in from not being null.
<!-- Reviewable comment -K-Ly-vxN5FNrcdY6JPW:57807386 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1213 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyJq69IPqw-JsbKiE)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1213)):</sup>
Do you actually use this? Seems like this is more of a one-time debugger type of line than anything that woudl ever be useful in debugging an error. Perhaps not, but it's another point of maintenance and was one more thing to update. Not sure if the value provided is worth it.
<!-- Reviewable comment -K-LyJq69IPqw-JsbKiF:370318212 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1218 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Lz4ROxTDW_9KAhyUv)** ([raw file](https://github.com/triplea-game/triplea/blob/6c1e6e60ea3fcb21b03472d8d68f836a5ddc0145/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1218)):</sup>
I support not returning an int, but void indicates we are using in/out parameters. That is a code smell. I'd return a modified resourceData object. This way the master method that calls this, can then keep track of the state of that object as it gets passed around. Instead you have to dig into this method to figure out what is going on.
<!-- Reviewable comment -K-Lz4ROxTDW_9KAhyUw:-988727432 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1271 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyaPk5aWnU4yl24qt)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1271)):</sup>
bad design here, we're relying on side effects. It's also unclear exactly what the "purchase" function should do.  To understand this code, you have to go look up the purchase method. that is quite a lot of effort to understand one line of code. Fix by trying to create a return type that has some meaning, and also by making the method name a bit more descriptive perhaps. From the context of this code in this method, this line is not understandable on its own.
<!-- Reviewable comment -K-LyaPk5aWnU4yl24qu:759833342 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1348 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LzS6qC1VIcR0i3cs2)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1348)):</sup>
again, here, returning resourceData, even if it is the same but modified resourceData that was passed in would make things more clear.
<!-- Reviewable comment -K-LzS6qC1VIcR0i3cs3:597937581 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProResourceData.java, line 42 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LxdNtVgZbZ0NnbuYN)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProResourceData.java#L42)):</sup>
confirmTempPurchases is a smell, violation of data encapsulation. A caller that calls this code will need to know exactly *when*, *why*, and what effect this method call will have. 
<!-- Reviewable comment -K-LxdNtVgZbZ0NnbuYO:253351407 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/205)
<!-- Sent from Reviewable.io -->

DanVanAtta commented:
Overall, I think its a step in the right direction, but not as clean as it could be. Particularly the "programming by side effects" elements. Extracting existing code into testable modules and then modifying those with a resourceData object IMO would be a good way to go to. The private methods look to be easily extractable concise objects that you could test.

As is, it is hard to say if it correctly fixes everything, and you may have some NPE possibilities.

DanVanAtta commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/205)
<!-- Sent from Reviewable.io -->

ron-murhammer commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1210 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Ly-vxN5FNrcdY6JPV)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1210)):</sup>
Not following you here? resourceData is used in most of the methods...
<!-- Reviewable comment -K-M7zrlzmQ7HyF5yk6w:1260206480 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1213 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyJq69IPqw-JsbKiE)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1213)):</sup>
So the AI log is useful and knowing how many resources the AI still has to use if important in following the methods.
<!-- Reviewable comment -K-M8PV5zbJxruJgbWWX:2006854667 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1218 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Lz4ROxTDW_9KAhyUv)** ([raw file](https://github.com/triplea-game/triplea/blob/6c1e6e60ea3fcb21b03472d8d68f836a5ddc0145/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1218)):</sup>
So I could return the resourceData object but its a member variable so don't really see the point.
<!-- Reviewable comment -K-M8XEvHEz5EidZ4BOO:1575610834 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1271 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyaPk5aWnU4yl24qt)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1271)):</sup>
I'm open to suggestions for a better method name. I personally think it makes sense as you are purchasing the unit but since I'm close to the code I'm probably biased.
<!-- Reviewable comment -K-MH79UcqcQDrIoPYOP:2013758077 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1348 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LzS6qC1VIcR0i3cs2)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1348)):</sup>
So resourceData isn't changed by that method. 
<!-- Reviewable comment -K-MHK2YRywRk6g4d-u5:-1789582841 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProResourceData.java, line 42 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LxdNtVgZbZ0NnbuYN)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProResourceData.java#L42)):</sup>
Umm... I don't particularly like it but couldn't think of a better way to do it. Essentially, it accumulates temp purchases then either confirms them or clears them.
<!-- Reviewable comment -K-MI9p5H-NTWkohwa6X:-45023941 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java, line 132 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-FuYWEmZKiNRy63PYl-r1-132)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java#L132)):</sup>
I can changed it to purchaseOption if you think that's better? Open to other suggestions.
<!-- Reviewable comment -K-MIdu30HpMil_YxKRQ:697007148 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/205)
<!-- Sent from Reviewable.io -->

ron-murhammer commented:
Sorry about the merge conflicts. This is why I wanted to get all the formatting stuff done at one time as commits that touch lots of files end up conflicting with all other PRs then keep having to be updated.

Honestly, I didn't do much refactoring with this change. Mostly just replacing 'int PUs' with a resource data object to track all resources. There is a ton of refactoring the needs done to clean this class up in general but just way beyond the scope here. 
<!-- Reviewable comment -K-MJVenngMpgLHltpO9:631271499 -->

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/205)
<!-- Sent from Reviewable.io -->

DanVanAtta commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1210 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Ly-vxN5FNrcdY6JPV)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1210)):</sup>
If passing `resourceData`  around as a parameter would be painful, can it be made immutable?
<!-- Reviewable comment -K-MXrgw4P_ptWCg9g7L:600180321 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1213 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyJq69IPqw-JsbKiE)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1213)):</sup>
Trace logging is generally a code smell. There are a number of interesting reads, some very quick googling has some interesting content: http://www.lenholgate.com/blog/2005/07/no-thats-not-the-point-and-yes-trace-logging-is-bad.html

I can agree with the log being useful. I recommend creating an "AILogger" class that can wrap this call to LogUtils. You can expand the usage of it as you come across/update other code. Such a utility class would make it easy to fine everywhere you have business/application level logging, and it would encourage DRY, and also modularity of how you handle application logging.
<!-- Reviewable comment -K-MYLYtatdvqpcX4Bg_:-1390835254 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1218 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Lz4ROxTDW_9KAhyUv)** ([raw file](https://github.com/triplea-game/triplea/blob/6c1e6e60ea3fcb21b03472d8d68f836a5ddc0145/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1218)):</sup>
Agreed, an updated method name perhaps at least would help. Part of the issue is that the "level" of abstraction is about the same between this helper method and the method it is helping. Breaking up the method it is helping a bit more, then perhaps it'll be easier to contrast this method name against others, and it'll make more sense. ie: doPurchaseSeaUnit, doPurchaseLandUnits.. etc.. 
<!-- Reviewable comment -K-Ma-OiPTWZ5z3hrCK5:-1897879750 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1271 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyaPk5aWnU4yl24qt)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1271)):</sup>
A method rename would only go so far. Breaking up the logic in this method into other static methods and/or possibly other classes when you need to  track state I think would be more appropriate. Making resourceData immutable should help too.
<!-- Reviewable comment -K-MZgC5iXvRIyh_36z4:717663428 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1348 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LzS6qC1VIcR0i3cs2)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1348)):</sup>
There is a side effect that occurs after calling this method, otherwise this line of code would not be here. Exactly what that is, or which parameters are in/out, I could not tell you. I'd have to spend more time with the method to understand what it does. So the 7 parameter method call is useful for following DRY, but it does not seem to be easily understood. Can the method be broken up into intermediary steps with fewer parameters maybe? There should be a number of ways to fix this situation.
<!-- Reviewable comment -K-Matlqx9mWsmMeCZ9P:1136161543 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 2135 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-McS5K1t5eBmgOxjGJ)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L2135)):</sup>
Log statement inside of a for loop. Looks like trace logging so I recommend removing. Was once useful, now no longer that you have tested this code and have made sure it works. If there is a need for this log statement, creating a string buffer before the loop and appending to it and logging once is recommended.
<!-- Reviewable comment -K-McS5K1t5eBmgOxjGK:-1427218564 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 2159 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Mczb5KT4xZ8jTDdnl)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L2159)):</sup>
This data structure is a good candidate for a first class object. Constructor could accept a `purchaseOptionsForTerritory` and the code below looks like it could be simplified. You could then test this object, it would be relatively simple. It would also give you a place to hide the magic numbers you have below, for example: `Math.pow(ppo.getCost(), 2))`. This value would be well assigned to a variable simply for the label of what it represents.  

Looking below, it seems like the exist point for the object would be the ProPurchaseOption selectOption. Perhaps convert that to:
`(new DefenseEfficiences(purchaseOptionsForTerritory)).randomizePurchaseOptions().getSelectionOption()`
<!-- Reviewable comment -K-Mczb6HmASLOXNorF4:208790842 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProResourceData.java, line 0 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LxdNtVgZbZ0NnbuYN)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProResourceData.java#L0)):</sup>
Splitting it to it's own object may help. Two objects that follow SRP could do the trick maybe. If one is dedicated to tracking/accumulating state, then I think that is pretty simply/easy.
<!-- Reviewable comment -K-Mdw6xpVxKNVO6f_QW:622640207 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java, line 0 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-FuYWEmZKiNRy63PYl-r1-132)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java#L0)):</sup>
that's seems fine. The obvious alternative is `proPurchaseOption`, the shorter version has my preference between the two.
<!-- Reviewable comment -K-MesJ08fUz9kTivcWX:1327288543 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/205)
<!-- Sent from Reviewable.io -->

DanVanAtta commented:
When #149 is merged, I think that'll be the end of the mass formatting. I was hoping for it to be done more swiftly as well, but the real world is messy I suppose.

#149 has been hanging for a while. Let's check with @Veqryn on his plans to review it. If he has any major changes maybe he can update that PR or add them on top so we can integrate it.

I can see desire to keep the scope small. But it does seem we adding a bit more to the mess than we are cleaning it. If this update was not sufficient reason to refactor it, I'm not sure what will be. Also, without reason to update it, a refactor project of this class is not terribly valuable unless we are very worried about regression bugs.

By that token, at least having any new code added be tested is a good start. Often decent sized modules can be created by extracted chunks of existing logic and then testing those new modules. It simplifies the existing code, since it now gets a layer of abstraction below it, and also makes it more robust since that layer is tested.

Another approach perhaps would be to just start trying to unit test the class from a high level. A high level test would be useful for basic regression cases, and it might make it clear where the largest value refactors would be. Not to mention the test would allow for a refactor to begin.
<!-- Reviewable comment -K-Mfae3oF49UZoLdrfT:-1771981154 -->

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/205)
<!-- Sent from Reviewable.io -->

ron-murhammer commented:
Yeah so I think we can merge #149.

Refactoring this class would quickly spill over to refactoring many related classes. It needs done this PR is focused on an enhancement. Unit testing is an interesting topic around AI code. It can be challenging to mock the data out and also to take into account the variability.
<!-- Reviewable comment -K-PTQxMmOTLoFb6576Z:2145934325 -->

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1210 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Ly-vxN5FNrcdY6JPV)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1210)):</sup>
It can't be immutable. It is essentially being used to track how many resources the AI has left as it plans its purchases. Previously it was just an 'int remainingPUs' which only took into account PUs that AI had. Now its an object that tracks all resources and as purchases are made it decrements the remaining resources.
<!-- Reviewable comment -K-POggvl8wIQowEqRUF:-366178036 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1213 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyJq69IPqw-JsbKiE)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1213)):</sup>
So AI logging is actually very useful and there is a window you can open in game to view the AI thinking over its moves. This is used mostly as a debugging tool as trying to figure out why the AI moved somewhere or purchased something given a particular scenario is very difficult without some sort of logging. The LogUtils class is the AI logging class (in AI packages and only used there). I think you are just misunderstanding the scope/purpose of LogUtils.
<!-- Reviewable comment -K-PP8BPZc4ySurdq9NT:-1772111880 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1218 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Lz4ROxTDW_9KAhyUv)** ([raw file](https://github.com/triplea-game/triplea/blob/6c1e6e60ea3fcb21b03472d8d68f836a5ddc0145/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1218)):</sup>
Agree. Just out of scope for this change.
<!-- Reviewable comment -K-PPvrUryYY46qEP-9A:934122907 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1271 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyaPk5aWnU4yl24qt)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1271)):</sup>
I don't think you quite understand the purpose of resourceData. Maybe it should be named resourceTracker or something. It is keeping track of the remaining resources the AI has as it plans its purchases.
<!-- Reviewable comment -K-PQ68wXNEAXb8xe9Ft:373500400 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1348 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LzS6qC1VIcR0i3cs2)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1348)):</sup>
Well you could divide it into a few methods but each one would still have a number of parameters. The problem is really how the scope of some of these parameters is handled which ends up having them passed around in lots of places. It would require a major refactor to actually improve this much.
<!-- Reviewable comment -K-PQKMLxQ7XQzx7U-_6:1147497777 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 2135 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-McS5K1t5eBmgOxjGJ)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L2135)):</sup>
So AI logging is different IMO then general application/system logging. When debugging you need to understand how the AI got to its conclusions which means you need a certain level of information. This method is checking all territories that potentially need additional defense units so logs the territory being checked and then some intermediate steps along with the conclusion of whether units were added.
<!-- Reviewable comment -K-PQqtlspee0icU1WaG:2125154769 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 2159 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Mczb5KT4xZ8jTDdnl)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L2159)):</sup>
Agree that creating an object would clean up this code but this actually has nothing to do with the changes in this PR. So out of scope.
<!-- Reviewable comment -K-PRFMNQuGzts0Loz1S:1000376068 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProResourceData.java, line 42 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LxdNtVgZbZ0NnbuYN)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProResourceData.java#L42)):</sup>
I thought about doing this but decided against it. You could have 1 object track all confirmed purchases and 1 track all temp purchases. Then you'd need methods to interface between them. In the methods using them you'd have 2 objects floating around tracking resources as purchases are made then trying to have them interact as well. It ended up being way messier than having 1 object track everything and just having methods to clear/confirm purchases.
<!-- Reviewable comment -K-PSOpqMU5Ntw6lt_KG:780979988 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/205)
<!-- Sent from Reviewable.io -->

ron-murhammer commented:
So I ended up just quickly going over the formatting of all the AI classes and cleaning it up by hand. I would focus on reviewing the functional changes in the initial 3 commits as the last 2 commits are just merging in the formatting changes from master then hand formatting the classes.

DanVanAtta commented:
Review status: 2 of 25 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1210 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Ly-vxN5FNrcdY6JPV)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1210)):</sup>
Okay, failing immutable, can you make the reference to it final?
<!-- Reviewable comment -K-QVKqlkUQdemBxx5DV:1442693169 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1213 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyJq69IPqw-JsbKiE)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1213)):</sup>
LogUtils is not a good name for a AI logging utility class. It's obvious then why I misunderstand the scope and purpose of LogUtils - it has the wrong name. To boot, the extra parameter for log granularity is a violation of DRY. So create another method in the logger util for application level logging for AI so you don't have to redundantly specify that log level.

I understand AI logging is useful, but trace logging to tell you every method that the code is descending through is a stop-gap for not having tests and is a way to debug the code. It is a test once and never test again type of testing. Hence, the debug statements are just as good as System.out.println statements, should be removed since they cause more clutter than anything. But here I can see that this is an application level log and not trace. The mixutre of these two use cases is rampant, which is a violation of SRP as well. 
<!-- Reviewable comment -K-QVT39gp3LZcgBd6mC:-1424320213 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1218 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Lz4ROxTDW_9KAhyUv)** ([raw file](https://github.com/triplea-game/triplea/blob/6c1e6e60ea3fcb21b03472d8d68f836a5ddc0145/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1218)):</sup>
Grant it, this is a really nice band-aid you've added on top of the code, but it does seem like this is more a band-aid than a deep fix.
At some point one must do more than just band-aid the code. Wishing for a refactor project to come along and magically unit test and refactor all the code just never happens in reality. Either you refactor and test it before making a fix, or you do it after, or it doesn't get done.

On goal of CR's is not to gatekeep and make sure code is perfect before shipped, but it is to try and have *some* more time spent improving quality in any way. So really my goal has been to point out where/why there are a lot of issues that still remain. I don't expect you to be able to fix it all, but if you make any improvements this will have been worthwhile hopefully.
<!-- Reviewable comment -K-QWCAlBc9-pvA2OHYk:1828558873 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1271 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyaPk5aWnU4yl24qt)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1271)):</sup>
That small rename would help a little yes, I'd say it is worth it
<!-- Reviewable comment -K-QW2aMOIhmhD7uBk_D:-725540164 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 2135 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-McS5K1t5eBmgOxjGJ)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L2135)):</sup>
Sure, they may have different purposes, but this is clearly debug logging. It may as well be a System.out.. Not only that, it would pollute whatever log with one line for every territory, every time the AI is considering a purchase. This log statement should likely at least be combined with any inside of the body fo the for loop if any so you can have on the same line the result of considering the territory, and not have to look at the next line to know what happened. To boot, if you look at the next line you know that this code generated output in exactly that format, chances are good one is reviewing the output in parallel with having the code side by side. This is describing a code development debugging process, not operational troubleshooting. A hallmark of good operational logging is that you will never need to look at the code to understand what each log message is telling you.
<!-- Reviewable comment -K-QWwVlbkZCwExVWHVu:-437717597 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 2159 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Mczb5KT4xZ8jTDdnl)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L2159)):</sup>
It somewhat does, you're modified a method that is a few hundred lines long and there are problems left and right. There are lots of opportunities to break up the long method. The method has another problem, not only is it long but the levels of abstraction are confused and mixed from higher level abstraction to lower level.  
<!-- Reviewable comment -K-QXXA0RENckk-t2uKh:-1964493506 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProResourceData.java, line 0 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LxdNtVgZbZ0NnbuYN)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProResourceData.java#L0)):</sup>
Is there any way then to increase the level of functionality in the object? Have two private variables with getter/setters, they may as well be public variables, this causes the implementation details of this class to be leaked/coupled unnecessarily.  Said in another way, there is not much code in this class, can we move some of the functions form the long ProPurchase method to this class? Have it accept a player or territory object for example to do some computation. That would be better than letting ProPurchase AI manage the data state of this class.
<!-- Reviewable comment -K-QXqjKWf_95QqRRLuf:623228135 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/205)
<!-- Sent from Reviewable.io -->

DanVanAtta commented:
Addendum @ron-murhammer  -  It seems like the core of the changes are in place and will not change beyond renaming or small things. Those changes would be easier to review as their own diff. Kindly, from a reviewer perspective, I'd rather see a second PR with one commit and this one merged, rather than this PR gain an additional commit.

At the same time, enough time has spent between the two of us thinking how to improve this code yet some more, I'd recommend spending 10-15 minutes as a time box to do whatever you can to quickly improve things one last pass before you let this code sit again for some time. The alternative is to merge and move on, or to spend some *real* time refactoring/rewriting the code (which you've indicated is not in your current interests right now, which I can completely agree with)

ron-murhammer commented:
Alright. I'll merge this and do a few of the small clean up tasks in another PR. Though even if I just added a last commit to this you could always just review that commit rather than all of them combined.
<!-- Reviewable comment -K-Ql1xGawVHjUeCmWl0:-1701448895 -->

---

Review status: 2 of 25 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1210 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Ly-vxN5FNrcdY6JPV)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1210)):</sup>
Can't as its a member variable that is instantiated each new purchase round.
<!-- Reviewable comment -K-Q_CC2Qq9T7xaVn56f:1400992686 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 1213 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LyJq69IPqw-JsbKiE)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L1213)):</sup>
So the AI log actually does use different log levels to organize information in a more readable way and allow filtering. The AI logging actually really has nothing to do with unit testing. There needs to be a way to have a log of decision points and data around those that the AI makes when running a game. Essentially be able to monitor intermediate decisions and data that lead to the eventual moves/purchases the AI makes. Otherwise trying to figure out why the AI made a dumb purchase or a dumb move is close to impossible. Again I encourage you to start up a game and have the AI play a turn then look at the log file window at  Game -> Show Hard AI Logs.
<!-- Reviewable comment -K-Q_VlqSKuH-O31xL30:404786115 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 2135 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-McS5K1t5eBmgOxjGJ)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L2135)):</sup>
I think there is just some misunderstanding still around the log intent. The purpose it to identify that the AI is checking defenses for a certain territory. It then will log subsequent lines around the decisions/data it computes then a final decision is logged. So this is sort of a scope log statement to identify what territory is being checked rather than having to put it at the beginning of each log statement.
<!-- Reviewable comment -K-QjZSjHWKbVJ-AUGw2:426796759 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java, line 2159 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-Mczb5KT4xZ8jTDdnl)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProPurchaseAI.java#L2159)):</sup>
So I did modify the method but not these lines. I modified something like 20 methods and just don't want to modify all of them at this point.
<!-- Reviewable comment -K-Qk7AIPhZ-IoyIDcFa:957376393 -->

---

<sup>**[src/games/strategy/triplea/ai/proAI/ProResourceData.java, line 42 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/205#-K-LxdNtVgZbZ0NnbuYN)** ([raw file](https://github.com/triplea-game/triplea/blob/6bfabbbb4a68828a50a13dddd2e4c65dba4378ff/src/games/strategy/triplea/ai/proAI/ProResourceData.java#L42)):</sup>
Not without refactoring essentially the entire ProPurchase class. The resourceData is used in the majority of the methods in it as it tracks what resources are remaining as the various purchase decision are made. You essentially have to add most of the code from ProPurchase into this object which would just end up with this being the giant class instead of that.
<!-- Reviewable comment -K-QkO2wicLoiTuiko-m:-231640640 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/205)
<!-- Sent from Reviewable.io -->

DanVanAtta commented:
Indeed you can see the per-commit diff - but if you comment on it then the comments are associated with the diff and not the pull request. In that case nobody is notified via email of the comment, and someone would have to just happen upon the commit comment to see it.

The other goal is that diffs for reviewers we want to be clean. I've certainly have spent some quality time just trying to clean up diffs to make things easier to review. Turns out writing code for it to be easy to review is more and more important than it was historically. I've been learning this more and more recently.

DanVanAtta commented:
w.r.t to logs, there is less disagreement than you might think. But having an application level logger be called LogUtils is not an easily understood design. Create some better abstractions, specific names to convey the exact intent. 

What does logLevel.FINE vs FINER in the AI log mean semantically? There is no way to tell without looking at a large sampling of LogUtils calls with those two log levels. Would be preferable to see something like this:

```
AILogger.debug("");
AILogger.log(""); 
```

instead of:
```
LogUtils.log(LogLevel.FINER, "");
LogUtils.log(LogLevel.FINE, "");
```

One clearly conveys the intent and purpose, the other does not. Hence, when the latter is used as an AI logger, it is confusing.
